backends/tc/ebpfCodeGen.cpp

```diff
@@ -407,14 +407,170 @@ void TCIngressPipelinePNA::emitLocalVariables(EBPF::CodeBuilder *builder) {
+            type->emit(builder);
+            builder->appendFormat(", %d)", widthToExtract);
+        }
+
+        builder->append(")");
+        builder->endOfStatement(true);
+    } else {
+        if (program->options.arch == "psa" && widthToExtract % 8 != 0) {
+            // To explain the problem in error lets assume that we have bit<68> field with value:
+            //   0x11223344556677889
+            //                     ^ this digit will be parsed into a half of byte
+            // Such fields are parsed into a table of bytes in network byte order, so possible
+            // values in dataplane are (note the position of additional '0' at the end):
+            //   0x112233445566778809
+            //   0x112233445566778890
+            // To correctly insert that padding, the length of field must be known, but tools like
+            // nikss-ctl (and the nikss library) don't consume P4info.txt to have such knowledge.
+            // There is also a bug in (de)parser causing such fields to be deparsed incorrectly.
+            ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET,
+                    "%1%: fields wider than 64 bits must have a size multiple of 8 bits (1 byte) "
+                    "due to ambiguous padding in the LSB byte when the condition is not met",
+                    field);
+        }
+
+        // wide values; read all bytes one by one.
+        unsigned shift;
+        if (alignment == 0)
+            shift = 0;
+        else
+            shift = 8 - alignment;
+
+        const char *helper;
+        if (shift == 0)
+            helper = "load_byte";
+        else
+            helper = "load_half";
+        auto bt = EBPF::EBPFTypeFactory::instance->create(IR::Type_Bits::get(8));
+        unsigned bytes = ROUNDUP(widthToExtract, 8);
+        for (unsigned i = 0; i < bytes; i++) {
+            builder->emitIndent();
+            visit(expr);
+            builder->appendFormat(".%s[%d] = (", fieldName.c_str(), i);
+            bt->emit(builder);
+            builder->appendFormat(")((%s(%s, BYTES(%s) + %d) >> %d)", helper,
+                                  program->packetStartVar.c_str(), program->offsetVar.c_str(), i,
+                                  shift);
+
+            if ((i == bytes - 1) && (widthToExtract % 8 != 0)) {
+                builder->append(" & EBPF_MASK(");
+                bt->emit(builder);
+                builder->appendFormat(", %d)", widthToExtract % 8);
+            }
+
+            builder->append(")");
+            builder->endOfStatement(true);
+        }
+    }
+
+    builder->emitIndent();
+    builder->appendFormat("%s += %d", program->offsetVar.c_str(), widthToExtract);
+    builder->endOfStatement(true);
+
+    // eBPF can pass 64 bits of data as one argument passed in 64 bit register,
+    // so value of the field is printed only when it fits into that register
+    if (widthToExtract <= 64) {
+        cstring exprStr = expr->is<IR::PathExpression>()
```

Comment:
I believe you can use `toString` in both cases here. The outcome is the same.

Reply:
Addressed.